### ISSUE TITLE
Inconsistent Docker Image Name between docker-hub trigger and manual run

### ISSUE THREAD

**Issue by alexis2b** [spinnaker/spinnaker#4982]
### Issue Summary:
The same Docker Image is sent to the pipeline with different names when using an automated trigger vs. a manual run. This prevents successful binding of the image to the Deployment Manifest working for both scenarios.

### Cloud Provider(s):
Kubernetes Provider V2

### Environment:
Spinnaker 1.16.1

### Feature Area:
Docker Registry auto-trigger vs. Deck manual run

### Description:
The "Docker Registry" automated trigger type and the "Manual Run" from Deck put an inconsistent artifact name in the trigger payload:
- Docker Registry will put `"name": "index.docker.io/alexisdb/my-app"`
- Deck Manual run will put `"name": "alexisdb/my-app"` (i.e. the `index.docker.io/` prefix is not there)

This is a problem when trying to bind the Docker Image with the Manifest to deploys in the Deploy (Manifest) stage. If the manifest contains `index.docker.io/alexisdb/my-app` as the image name in the Deployment resource, it will fail to be bound with the name `alexisdb/my-app` when running manually (and vice-versa).

### Steps to Reproduce:
Create a simple pipeline with:
- Automated Trigger from Docker Hub
- Expected artifact: the docker image from the trigger
- Deploy (Manifest) stage with the image name + "bind required artifacts" set to Docker image

Kickoff automatically when a new image is uploaded.
Kickoff manually from Deck.

Since the image name will be different, either the Expected Artifact match will fail (can be worked around with Regexps to make the "index.docker.io" prefix optional) but also the Deployment (Manifest) stage will fail because it expects the exact image name (and I could not find a workaround here).

### Additional Details:

JSON trigger when automated trigger kicks in:
```
    "trigger": {
        "type": "docker",
        "user": "[anonymous]",
        "parameters": {},
        "artifacts": [
            {
                "customKind": false,
                "reference": "index.docker.io/alexisdb/my-app:1.0.22",
                "name": "index.docker.io/alexisdb/my-app",
                "type": "docker/image",
                "version": "1.0.22"
            }
        ],
        "notifications": [],
        "account": "dockerhub",
        "repository": "alexisdb/my-app",
        "tag": "1.0.22",
        "resolvedExpectedArtifacts": [
            {
                "usePriorArtifact": false,
                "useDefaultArtifact": false,
                "defaultArtifact": {
                    "customKind": true,
                    "metadata": {
                        "id": "d5508ee7-0b81-432f-ba3c-cff01b068dee"
                    }
                },
                "id": "a9d0d761-9bfc-4a9e-b372-d2d424449c54",
                "boundArtifact": {
                    "customKind": false,
                    "reference": "index.docker.io/alexisdb/my-app:1.0.22",
                    "name": "index.docker.io/alexisdb/my-app",
                    "type": "docker/image",
                    "version": "1.0.22"
                },
                "matchArtifact": {
                    "customKind": false,
                    "metadata": {
                        "id": "3edd7545-5b31-4f29-8a0d-897ebeb607a4"
                    },
                    "name": "(index.docker.io/)?alexisdb/my-app",
                    "artifactAccount": "docker-registry",
                    "type": "docker/image"
                }
            }
        ],
        "rebake": false,
        "strategy": false,
        "dryRun": false,
        "enabled": true,
        "id": "48e4eee2-8790-376e-9ec0-16f399249e92",
        "expectedArtifacts": [ ... ],
        "eventId": "81066114-9f2b-4aee-825d-44364e2c3622"
    },
```

versus the JSON trigger when using Deck Manual Run and selecting the same tag:
```
    "trigger": {
        "type": "manual",
        "user": "[anonymous]",
        "parameters": {},
        "artifacts": [
            {
                "customKind": false,
                "reference": "alexisdb/my-app:1.0.34",
                "name": "alexisdb/my-app",
                "type": "docker/image",
                "version": "1.0.34"
            }
        ],
        "notifications": [],
        "account": "dockerhub",
        "repository": "alexisdb/my-app",
        "tag": "1.0.34",
        "resolvedExpectedArtifacts": [
            {
                "usePriorArtifact": false,
                "useDefaultArtifact": false,
                "defaultArtifact": {
                    "customKind": true,
                    "metadata": {
                        "id": "f7372d70-9206-4dfd-b8bd-f1214027deb9"
                    }
                },
                "id": "775a8d65-9bc5-4168-a177-67e77390adda",
                "boundArtifact": {
                    "customKind": false,
                    "reference": "alexisdb/my-app:1.0.34",
                    "name": "alexisdb/my-app",
                    "type": "docker/image",
                    "version": "1.0.34"
                },
                "matchArtifact": {
                    "customKind": false,
                    "metadata": {
                        "id": "02de4773-ce6c-49cf-bc9b-06056227df0a"
                    },
                    "name": "(index.docker.io/)?alexisdb/my-app",
                    "artifactAccount": "docker-registry",
                    "type": "docker/image"
                }
            }
        ],
        "rebake": false,
        "strategy": false,
        "dryRun": false,
        "enabled": true,
        "id": "7571678e-66ad-381f-b68b-4ffc7d6d98ab",
        "expectedArtifacts": [...],
        "eventId": "b8496522-69f7-4edb-acba-b012bcdb8a74",
        "executionId": "01DP3W75CD9P08S5PEMZ3X0701"
    },
```

As you can see the value or trigger.artifacts.name is different, although it is the same image.

**Comment by alexis2b** [comment]
Seems related to #4955 and possibly also to #4956

**Comment by maggieneterval** [comment]
Hey @alexis2b, thanks for the detailed issue report! Would you mind posting a sanitized version of your pipeline JSON? Specifically, I'm interested in your pipeline's `expectedArtifacts` and `triggers`. Thanks!

**Comment by alexis2b** [comment]
Hi @maggieneterval, thanks for the quick reply!
Here is the mostly untouched JSON of the pipeline - I removed the notifications and changed the docker account name to 'alexisdb', everything else is the same.
```
{
    "expectedArtifacts": [
      {
        "defaultArtifact": {
          "customKind": true,
          "id": "f7372d70-9206-4dfd-b8bd-f1214027deb9"
        },
        "displayName": "my-app-docker-image",
        "id": "775a8d65-9bc5-4168-a177-67e77390adda",
        "matchArtifact": {
          "artifactAccount": "docker-registry",
          "id": "02de4773-ce6c-49cf-bc9b-06056227df0a",
          "name": "(index.docker.io/)?alexisdb/my-app",
          "type": "docker/image"
        },
        "useDefaultArtifact": false,
        "usePriorArtifact": false
      },
      {
        "defaultArtifact": {
          "artifactAccount": "no-auth-http-account",
          "customKind": true,
          "id": "fff41984-088e-4fc9-b0be-fe101460f09b",
          "reference": "https://nexus/repository/spinnaker-test/my-app/my-app-${trigger.tag}.tgz",
          "type": "http/file"
        },
        "displayName": "my-app-helm-chart",
        "id": "5fd40d7c-4011-42ca-a7a1-2a05e9650afb",
        "matchArtifact": {
          "artifactAccount": "no-auth-http-account",
          "customKind": true,
          "id": "8ae2ab44-dd9f-47ef-bd1e-821219d89ad5",
          "reference": "https://nexus/repository/spinnaker-test/my-app/my-app.tgz",
          "type": "http/file"
        },
        "useDefaultArtifact": true,
        "usePriorArtifact": false
      },
      {
        "defaultArtifact": {
          "artifactAccount": "no-auth-http-account",
          "customKind": true,
          "id": "e5c3d3f4-a2a2-44f5-b0f2-fabc7b7a35cc",
          "reference": "https://nexus/repository/spinnaker-test/my-app/my-app-values-integration.yaml",
          "type": "http/file"
        },
        "displayName": "my-app-values-integration",
        "id": "07a2b380-2e1b-4633-a8ca-51ac04356b50",
        "matchArtifact": {
          "artifactAccount": "no-auth-http-account",
          "customKind": true,
          "id": "b701c0f6-9c4e-4515-ba9f-665ec17f44b1",
          "reference": "/my-app/my-app-values-integration.yaml",
          "type": "http/file"
        },
        "useDefaultArtifact": true,
        "usePriorArtifact": false
      }
    ],
    "keepWaitingPipelines": false,
    "lastModifiedBy": "anonymous",
    "limitConcurrent": true,
    "notifications": [],
    "stages": [
      {
        "expectedArtifacts": [
          {
            "defaultArtifact": {
              "customKind": true,
              "id": "32fd0ef5-4ed0-4eaa-bc79-0433e0278dea"
            },
            "displayName": "my-app-manifests",
            "id": "820d7596-fc97-4704-b59b-3b1312393359",
            "matchArtifact": {
              "id": "5cc93257-7804-4f36-96b2-2f320ad4c284",
              "name": "my-app-manifests",
              "type": "embedded/base64"
            },
            "useDefaultArtifact": false,
            "usePriorArtifact": false
          }
        ],
        "failOnFailedExpressions": true,
        "inputArtifacts": [
          {
            "account": "no-auth-http-account",
            "id": "5fd40d7c-4011-42ca-a7a1-2a05e9650afb"
          },
          {
            "account": "no-auth-http-account",
            "id": "07a2b380-2e1b-4633-a8ca-51ac04356b50"
          }
        ],
        "name": "Bake (Manifest)",
        "namespace": "spin-integration",
        "notifications": [],
        "outputName": "my-app",
        "overrides": {},
        "refId": "1",
        "requisiteStageRefIds": [],
        "sendNotifications": true,
        "templateRenderer": "HELM2",
        "type": "bakeManifest"
      },
      {
        "account": "integration",
        "cloudProvider": "kubernetes",
        "failOnFailedExpressions": true,
        "manifestArtifactAccount": "embedded-artifact",
        "manifestArtifactId": "820d7596-fc97-4704-b59b-3b1312393359",
        "moniker": {
          "app": "my-app"
        },
        "name": "Deploy (Manifest)",
        "notifications": [],
        "refId": "2",
        "requiredArtifactIds": [
          "775a8d65-9bc5-4168-a177-67e77390adda"
        ],
        "requiredArtifacts": [],
        "requisiteStageRefIds": [
          "1"
        ],
        "sendNotifications": true,
        "skipExpressionEvaluation": false,
        "source": "artifact",
        "trafficManagement": {
          "enabled": false,
          "options": {
            "enableTraffic": false,
            "services": []
          }
        },
        "type": "deployManifest"
      },
      {
        "name": "Wait",
        "refId": "3",
        "requisiteStageRefIds": [
          "2"
        ],
        "type": "wait",
        "waitTime": 5
      }
    ],
    "triggers": [
      {
        "account": "dockerhub",
        "enabled": true,
        "organization": "alexisdb",
        "registry": "index.docker.io",
        "repository": "alexisdb/my-app",
        "type": "docker"
      }
    ],
    "updateTs": "1569940795000"
  }
```

Also here is what the key parts of the Deployment resource looks like before the docker image is bound:
```
# Source: my-app/templates/deployment.yaml
apiVersion: apps/v1
kind: Deployment
...
spec:
...
  template:
...
    spec:
      containers:
        - name: my-app
          image: index.docker.io/alexisdb/my-app
          imagePullPolicy: Always
```

which gets replaced by the Deploy (Manifest) stage with the bound Docker Image into:
```
apiVersion: extensions/v1beta1
kind: Deployment
...
spec:
...
  template:
...
    spec:
      containers:
        - image: 'index.docker.io/alexisdb/my-app:1.0.36'
          imagePullPolicy: Always
```

The problem is when executing the same pipeline with Manual Run and selecting the right image + tag in the helpful dialog, the artifact name is `alexisdb/my-app` which fails to Bind to the expected name in the input Manifest.

**Comment by maggieneterval** [comment]
Thank *you* for the speedy reply as well! I think I've discovered the root cause of this bug and have submitted a patch to Spinnaker's Echo service to resolve it. Hopefully we'll be able to cherry-pick the fix into the next patch release of 1.16, I'll let you know if not. Cheers!
